relay/temp-pallets/pallet-staking-fixer/src/lib.rs

```diff
@@ -0,0 +1,126 @@
+//! `Staking::restore_ledger` call, providing sensible defaults that are enough to restore all
+//! corrupted ledgers.
+//! - Extra safety check: if the resulting restored ledger from `Staking::restore_ledger` has a
+//! larger total staked than the stash's free balance, the ledger is unstaked.
+//!
+//! Note: this pallet is temporary and can be removed when all the ledgers in Polkadot and Kusama
+//! are restored.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::traits::Currency;
+use frame_system::ensure_signed;
+use pallet_staking::WeightInfo;
+use sp_staking::StakingAccount;
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::{pallet_prelude::*, RawOrigin};
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_staking::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Ledger of `stash` has been recovered.
+		Restored { stash: T::AccountId },
+		/// Ledger of `stash` has been recovered. The resulting recoving ended up in unbonding and
+		/// the ledger to unstake.
+		RestoredUnstaked { stash: T::AccountId },
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Restores the state of a staking ledger that is in an inconsistent state.
+		///
+		/// This call functions as a proxy to `Staking::restore_ledger`, but exposes it to signed
+		/// origins. In addition, this call checks if, after the ledger has been successfully
+		/// restored, the ledger locks are higher or equal than the stash's free balance. If not, it
+		/// forces the unstake of the ledger.
+		///
+		/// Safety note: Only ledgers associated `stash` that are corrupted will be mutated. Thus it
+		/// is safe to expose `Staking::restore_ledger` through the signed origin.
+		#[pallet::call_index(0)]
+		#[pallet::weight(
+            <T::WeightInfo>::restore_ledger() +
+            <T::WeightInfo>::force_unstake(maybe_slashing_spans.unwrap_or_default())
+        )]
+		pub fn restore_ledger(
+			origin: OriginFor<T>,
+			stash: T::AccountId,
+			maybe_slashing_spans: Option<u32>,
+		) -> DispatchResultWithPostInfo {
+			let _ = ensure_signed(origin)?;
+
+			// calls `Staking::restore_ledger` as `Root`.
+			pallet_staking::Pallet::<T>::restore_ledger(
+				RawOrigin::Root.into(),
+				stash.clone(),
+				None,
+				None,
+				None,
+			)?;
+
+			let ledger = pallet_staking::Pallet::<T>::ledger(StakingAccount::Stash(stash.clone()))?;
+
+			// check if stash's free balance covers the current ledger's total amount. If not,
+			// force unstake the ledger.
+			let weight = if ledger.total > T::Currency::free_balance(&stash) {
```

Comment:
How many staking ledgers fall into this category? if it is a handful that have to be unstaked, I would rather we hardcode their accounts, rather than 'codify' their conditions, so as to not risk this code path being accidentally accessibly to any future state.

Reply:
The problem with hardcoding stashes for this check is that the condition may change with time from now until we call the extrinsic.

However, we do hardcode the stash accounts that need to be restored now. Through that, we can ensure that only the ledgers that 1. are corrupted and 2. are whitelisted can be recovered, which adds another layer of safety. The whitelisted accounts are added in the runtime configs.